src/condor_credd/condor_credmon_oauth/condor_credmon_oauth

```diff
@@ -80,22 +94,68 @@ def start_credmon_process(q, cred_dir, args):
-    ]
+    # Each provider needs to "stay in their lane" and be told which service providers to use.
+    # - Administrators must spell out the provider configuration for the OAuth2 and Local credmons.
+    # - Users can specify arbitrary provider names and configurations for the Vault credmon
+    # For backward compat, if the providers aren't explicitly set, we use the following algorithm:
+    # - If it's `scitokens`, we assume it is a local credmon.
+    # - If config knob `FOO_CLIENT_ID` is non-empty, then we assume it's the OAuth2 credmon.
+    # - Otherwise, it's the Vault credmon's.
+
+    # The local credmon historically had a single provider; the Vault and OAuth credmon assumed they provided all tokens.
+    local_provider_names = set(filter(None, re.compile(r"[\s,]+").split(htcondor.param.get("LOCAL_CREDMON_PROVIDER_NAMES", ""))))
+    # Backward compat: if not providing a list, fallback to the singular config value.
+    if not local_provider_names:
+        local_provider_names = {htcondor.param.get("LOCAL_CREDMON_PROVIDER_NAME", "scitokens")}
+    oauth2_providers = set(filter(None, re.compile(r"[\s,]+").split(htcondor.param.get("OAUTH2_CREDMON_PROVIDER_NAMES", "*"))))
+    client_providers = set(filter(None, re.compile(r"[\s,]+").split(htcondor.param.get("CLIENT_CREDMON_PROVIDER_NAMES", ""))))
+
+    credmons = \
+        [OAuthCredmon(providers=oauth2_providers, cred_dir=cred_dir, args=args)] + \
+        [LocalCredmon(provider=provider, cred_dir=cred_dir, args=args) for provider in local_provider_names] + \
+        [ClientCredmon(provider=provider, cred_dir=cred_dir, args=args) for provider in client_providers]
+
+    vault_providers = set(filter(None, re.compile(r"[\s,]+").split(htcondor.param.get("VAULT_CREDMON_PROVIDER_NAMES", "*"))))
+    if VaultCredmon is None and vault_providers and vault_providers != {"*"}:
+        logger.warning("Vault credmon is configured via `VAULT_CREDMON_PROVIDER_NAMES` but the Vault credmon is not installed")
+        vault_providers = set()
+    elif VaultCredmon is not None:
+        credmons.append(VaultCredmon(providers=vault_providers, cred_dir=cred_dir, args=args))
+
+    # Ugly n^2 check for overlapping providers
+    providers = {
+        "local": local_provider_names,
+        "oauth2": oauth2_providers,
+        "client": client_providers,
+        "vault": vault_providers,
+    }
+    overlaps_found = set()
+    for provider1, s1 in providers.items():
+        if not s1:
+            continue
+        for provider2, s2 in providers.items():
+            if (not s2) or (provider1 == provider2) or ((provider2, provider1,) in overlaps_found) or not (s1 & s2):
+                continue
+            logger.warning(f"Overlaps found between {provider1.upper()} and {provider2.upper()} CREDMON_PROVIDER_NAMES: {', '.join(s1 & s2)}")
+            overlaps_found.add((provider1, provider2,))
+    if overlaps_found:
+        logger.warning("Continuing despite overlaps found in providers...")
```

Comment:
Assuming Python dictionary hashes have constant-time look-up, the following will be O(n); it will report two collisions rather than three if the same name is in three provider lists*, but that should be OK for our purposes.
```
name_to_provider_map = {}
for provider, names in providers.items():
    for name in names:
        collision = name_to_provider_map.get(name)
        if collision is None:
            name_to_provider_map[name] = provider
        else:
            print(f"{name} is registered to both {collision} and {provider}")
```
*: That is, if "a" is in lists local, oauth2, and client, this will report the local-oauth2 and local-client collisions, but not the oauth2-client collision.  This could be remedied by storing a _list_ of provider-names in each entry, but computing the complexity for that made my head hurt.

I guess a list of provider-names for each name might produce prettier output, too, e.g. "a is in the lists x, y, and z".

Reply:
Improved in 866d438789

Reply:
Thanks for the suggestion! Improved in 866d438789662139287bdceda6d41ae81414f8f4